Libraries/LibWeb/ContentSecurityPolicy/Directives/DirectiveOperations.cpp

```diff
@@ -224,4 +229,353 @@ FlyString get_the_effective_directive_for_inline_checks(Directive::InlineType ty
+//            For example, *.example.com host-part matches www.example.com, but www.example.com does not host-part match *.example.com.
+[[nodiscard]] static MatchResult host_part_matches(StringView pattern, Optional<URL::Host> const& maybe_host)
+{
+    // 1. If host is not a domain, return "Does Not Match".
+    // Spec Note: A future version of this specification may allow literal IPv6 and IPv4 addresses, depending on usage and demand.
+    //            Given the weak security properties of IP addresses in relation to named hosts, however, authors are encouraged
+    //            to prefer the latter whenever possible.
+    if (!maybe_host.has_value())
+        return MatchResult::DoesNotMatch;
+
+    auto const& host = maybe_host.value();
+
+    if (!host.is_domain())
+        return MatchResult::DoesNotMatch;
+
+    // 2. If pattern is "*", return "Matches".
+    if (pattern == "*"sv)
+        return MatchResult::Matches;
+
+    VERIFY(host.has<String>());
+    auto host_string = host.get<String>();
+
+    // 3. If pattern starts with "*.":
+    if (pattern.starts_with("*."sv)) {
+        // 1. Let remaining be pattern with the leading U+002A (*) removed and ASCII lowercased.
+        auto remaining = MUST(Infra::to_ascii_lowercase(pattern.substring_view(1)));
+
+        // 2. If host to ASCII lowercase ends with remaining, then return "Matches".
+        auto lowercase_host = MUST(Infra::to_ascii_lowercase(host_string));
+        if (lowercase_host.ends_with_bytes(remaining))
+            return MatchResult::Matches;
+
+        // 3. Return "Does Not Match".
+        return MatchResult::DoesNotMatch;
+    }
+
+    // 4. If pattern is not an ASCII case-insensitive match for host, return "Does Not Match".
+    if (!Infra::is_ascii_case_insensitive_match(pattern, host_string))
+        return MatchResult::DoesNotMatch;
+
+    // 5. Return "Matches".
+    return MatchResult::Matches;
+}
+
+// https://w3c.github.io/webappsec-csp/#port-part-matches
+// An ASCII string input port-part matches URL url if a CSP source expression that contained the first as a port-part
+// could potentially match a URL containing the latter’s port and scheme. For example, "80" port-part matches
+// matches http://example.com.
+[[nodiscard]] static MatchResult port_part_matches(Optional<StringView> input, URL::URL const& url)
+{
+    // FIXME: 1. Assert: input is the empty string, "*", or a sequence of ASCII digits.
+
+    // 2. If input is equal to "*", return "Matches".
+    if (input == "*"sv)
+        return MatchResult::Matches;
+
+    // 3. Let normalizedInput be null if input is the empty string; otherwise input interpreted as decimal number.
+    Optional<u16> normalized_input;
+    if (input.has_value()) {
+        VERIFY(!input.value().is_empty());
+        auto maybe_port = AK::StringUtils::convert_to_uint<u16>(input.value(), TrimWhitespace::No);
```

Comment:
nit, but it reads a bit nicer to do `input.value().to_number<u16>(TrimWhitespace::No)` (which calls this under the hood)